internal/bundle/utils.go

```diff
@@ -116,25 +177,62 @@ func LoadBundleFromDisk(path, name string, bvc *bundle.VerificationConfig) (*bun
 	}
 }
 
-// SaveBundleToDisk saves the given raw bytes representing the bundle's content to disk
-func SaveBundleToDisk(path string, raw io.Reader) (string, error) {
+// SaveBundleToDisk persists a bundle to disk. Bundles are wrapped in a 'bundlePackage' in order
+// to support additional metadata (e.g. etag) that is not part of the original bundle.
+func SaveBundleToDisk(path string, rawBundle io.Reader, opts *SaveOptions) error {
 	if _, err := os.Stat(path); os.IsNotExist(err) {
 		err = os.MkdirAll(path, os.ModePerm)
 		if err != nil {
-			return "", err
+			return fmt.Errorf("failed to create bundle directory: %w", err)
 		}
 	}
 
-	if raw == nil {
-		return "", fmt.Errorf("no raw bundle bytes to persist to disk")
+	// supplying no bundle data is an error case
+	if rawBundle == nil {
+		return fmt.Errorf("no raw bundle bytes to persist to disk")
+	}
+
+	// create a temporary, intermediary file to write the bundle package to
+	tempBundlePackageDestination, err := os.CreateTemp(path, ".bundlePackage.json.*.tmp")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary bundle package file: %w", err)
+	}
+
+	var etag string
+	if opts != nil && opts.Etag != "" {
+		etag = opts.Etag
+	}
+
+	// write the bundle package to the intermediary file in parts, first the metadata, followed by the bundle data.
+	// this is done to avoid loading the bundle data into memory only to write it to disk.
+	_, err = tempBundlePackageDestination.WriteString(fmt.Sprintf(`{"etag":"%s","bundle":"`, etag))
+	if err != nil {
+		return fmt.Errorf("failed to write metadata to bundle package file: %w", err)
+	}
+	// write the bundle data to the intermediary file, base64 encoded so that it can be unmarshalled as []byte
+	enc := base64.NewEncoder(base64.StdEncoding, tempBundlePackageDestination)
+	_, err = io.Copy(enc, rawBundle)
+	if err != nil {
+		return fmt.Errorf("failed to write bundle data to bundle package file: %w", err)
+	}
+	err = enc.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close bundle data base64 encoder: %w", err)
+	}
+	// complete the json file having written the bundle data
+	_, err = tempBundlePackageDestination.WriteString(`"}`)
+	if err != nil {
+		return fmt.Errorf("failed to complete writing of JSON data to bundle package file: %w", err)
```

Comment:
This seems a bit hacky. I wonder if we can define the `bundlePackage` struct in a way where the `Bundle` field has a value that is a struct that implements `io.Reader`. We can then json encode `bundlePackage` and pass it to the `io.Copy` call. I've not tried this approach so not sure if this will work.